File: B. Medium/0707-design-linked-list/0707-design-linked-list.c
typedef struct Node {
    int val;
    struct Node* next;
} Node;

typedef struct {
    Node* head;
    int length;
} MyLinkedList;

MyLinkedList* myLinkedListCreate() {
    MyLinkedList* list = (MyLinkedList*)malloc(sizeof(MyLinkedList));
    if (list == NULL) {
        fprintf(stderr, "Linked List Memory Allocation Failed");
        return NULL;
    }
    list->head = NULL;
    list->length = 0;
    return list;
}

int myLinkedListGet(MyLinkedList* obj, int index) {
    if ((index < 0) || (index >= obj->length)) {
        return -1;
    }
    else {
        Node* currentNode = obj->head;
        for (int i = 0; i < index; i++) {
            currentNode = currentNode->next;
        }
        return currentNode->val;
    }
}

void myLinkedListAddAtHead(MyLinkedList* obj, int val) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    if (newNode == NULL) {
        fprintf(stderr, "Node Memory Allocation Failed");
        return;
    }
    newNode->val = val;
    newNode->next = obj->head;
    obj->head = newNode;
    obj->length++;
}

void myLinkedListAddAtTail(MyLinkedList* obj, int val) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    if (newNode == NULL) {
        fprintf(stderr, "Node Memory Allocation Failed");
        return;
    }
    newNode->val = val;
    newNode->next = NULL;

    if (obj->head == NULL) {
        obj->head = newNode;
    }
    else {
        Node* currentNode = obj->head;
        while(currentNode->next != NULL) {
            currentNode = currentNode->next;
        }
        currentNode->next = newNode;
    }
    obj->length++;
}

void myLinkedListAddAtIndex(MyLinkedList* obj, int index, int val) {
    if ((index < 0) || (index > obj->length)) {
        return;
    }
    else if (index == 0) {
        myLinkedListAddAtHead(obj, val);
    }
    else if (index == obj->length) {
        myLinkedListAddAtTail(obj, val);
    }
    else {
        Node* newNode = (Node*)malloc(sizeof(Node));
        if (newNode == NULL) {
            fprintf(stderr, "Node Memory Allocation Failed");
            return;
        }
        newNode->val = val;
        newNode->next = NULL;
        Node* currentNode = obj->head;
        for (int i = 0; i < (index - 1); i++) {
            currentNode = currentNode->next;
        }
        newNode->next = currentNode->next;
        currentNode->next = newNode;
        obj->length++;
    }
}

void myLinkedListDeleteAtIndex(MyLinkedList* obj, int index) {
    if ((index < 0) || (index >= obj->length)) {
        return;
    }
    Node* currentNode = obj->head;
    if (index == 0) {
        obj->head = currentNode->next;
        free(currentNode);
    }
    else {
        Node* previousNode = NULL;
        for (int i = 0; i < index; i++) {
            previousNode = currentNode;
            currentNode = currentNode->next;
        }
        previousNode->next = currentNode->next;
        free(currentNode);
    }
    obj->length--;
}

void myLinkedListFree(MyLinkedList* obj) {
    Node* currentNode = obj->head;
    while (currentNode != NULL) {
        Node* nextNode = currentNode->next;
        free(currentNode);
        currentNode = nextNode;
    }
    free(obj);
}

/**
 * Your MyLinkedList struct will be instantiated and called as such:
 * MyLinkedList* obj = myLinkedListCreate();
 * int param_1 = myLinkedListGet(obj, index);
 
 * myLinkedListAddAtHead(obj, val);
 
 * myLinkedListAddAtTail(obj, val);
 
 * myLinkedListAddAtIndex(obj, index, val);
 
 * myLinkedListDeleteAtIndex(obj, index);
 
 * myLinkedListFree(obj);
*/

File: Difficulty/B. Medium/0707-design-linked-list/0707-design-linked-list.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

class MyLinkedList:
    def __init__(self):
        self.head = None
        self.length = 0

    def get(self, index: int) -> int:
        if ((index >= self.length) or (index < 0)):
            return -1
        else:
            currentNode = self.head
            while (index):
                currentNode = currentNode.next
                index -= 1
            return currentNode.val

    def addAtHead(self, val: int) -> None:
        newNode = Node(val)
        newNode.next = self.head
        self.head = newNode
        self.length += 1

    def addAtTail(self, val: int) -> None:
        newNode = Node(val)
        if (self.head is None):
            self.head = newNode
        else:
            currentNode = self.head
            while (currentNode.next is not None):
                currentNode = currentNode.next
            currentNode.next = newNode
        self.length += 1

    def addAtIndex(self, index: int, val: int) -> None:
        if ((index < 0) or (index > self.length)):
            return 
        elif (index == 0):
            self.addAtHead(val)
        elif (index == self.length):
            self.addAtTail(val)
        else:
            newNode = Node(val)
            currentNode = self.head
            for _ in range(index - 1):
                currentNode = currentNode.next
            newNode.next = currentNode.next
            currentNode.next = newNode
            self.length += 1

    def deleteAtIndex(self, index: int) -> None:
        if ((index < 0) or (index >= self.length)):
            return 
        elif (index == 0):
            self.head = self.head.next
        else:
            currentNode = self.head
            for _ in range(index - 1):
                currentNode = currentNode.next
            currentNode.next = currentNode.next.next
        self.length -= 1

# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)